src/jupyterlabcontroller/models/domain/rspimage.py

```diff
@@ -311,21 +435,76 @@ def _replace_contents(self, images: Iterable[RSPImage]) -> None:
         for image in unresolved:
             if image.digest in self._by_digest:
                 other = self._by_digest[image.digest]
-                if other.is_possible_alias:
-                    self._unresolved_aliases[image.digest].append(image)
-                else:
-                    image.resolve_alias(other)
+                self._resolve_duplicate(image, other)
             else:
                 self._unresolved_aliases[image.digest].append(image)
                 self._by_digest[image.digest] = image
 
-        # Now, all images have been resolved where possible. Take a second
-        # pass and register all images by name and type.
+        # Now, all images have been resolved where possible. Take a third
+        # pass and register all images by type.
         for image in images:
-            self._by_tag_name[image.tag] = image
             self._by_type[image.image_type].append(image)
 
         # Sort all of the images by reverse order so the newest are first.
         # This allows all_images to return sorted order efficiently.
         for image_list in self._by_type.values():
             image_list.sort(reverse=True)
+
+    def _resolve_duplicate(self, new: RSPImage, old: RSPImage) -> None:
```

Comment:
Thank goodness you not only unravelled all of this but documented the application attempts.